### ISSUE TITLE
function-redefined false positive with type hints

### ISSUE THREAD

**Issue by MatthewScholefield** [PyCQA/pylint#4936]
### Bug description

```python
from typing import Callable
from functools import wraps

do_something: Callable[[], int]


def with_config(func):
    @wraps(func)
    def wrapper():
        return func("arg")
    return wrapper


@with_config
def do_something(arg: str) -> int:
    return len(arg)
```

### Command used

```shell
pylint -E example.py
```

### Pylint output

```shell
************* Module example
example.py:16:0: E0102: function already defined line 4 (function-redefined)
```

### Expected behavior

No output; the function is not being redefined because the top instance is not a definition — it is a type declaration.

### Pylint version

```shell
pylint 2.10.2
astroid 2.7.3
Python 3.9.5 (default, May 24 2021, 12:50:35) 
[GCC 11.1.0]
```
